### PR TITLE
Talker Agent Framework (M0, M1, M3)

### PR DESCRIPTION
**Pull request by styin**
## Summary
- **M0/M2**: Proto schema, generation pipeline, voice layer (Opus decoder, RealtimeSTT listener, turn detector, Kokoro TTS speaker)
- **M3**: Inference pipeline — prompt assembly, LLM streaming with cancellation, sentence boundary detection, soul container (tag extraction, emotion normalization, vocab rules), conservative playback accounting
- **CI**: Talker workflow (ruff lint/format, pytest, mypy) and proto lint (buf lint + breaking change detection)
- **192 tests** covering sentence detection edge cases, soul container tag processing, turn detector threading, and import smoke tests

## E2E Testing (via test harness, not committed to branch)

A dev-only test harness (`scripts/test_harness.py`) was used to validate the full pipeline without the Gateway. Results:

### Tested flows
- [x] **Text → LLM → pipeline → output** — Prompt formatting with persona (SOUL.md + IDENTITY.md), LLM streaming via SSE `/v1/completions`, sentence boundary detection, soul container tag extraction (emotions), conversation history across turns, ResponseStarted/SentenceEvent/EmotionEvent/ResponseComplete sequencing
- [x] **Mic → STT → LLM → pipeline → output** — WSLg PulseAudio mic capture, VAD (WebRTC + Silero) speech detection, Whisper transcription (distil-large-v3), final transcript → TalkerContext → ProcessPrompt
- [x] **Mic → STT → LLM → pipeline → TTS → speaker** — Kokoro TTS synthesis, audio playback via PulseAudio/WSLg, multi-sentence playback
- [x] **SSE parser** — Fixed to read OpenAI-compatible `choices[0].text` format (was reading non-standard `content` field, silently dropping all tokens)
- [x] **Think block stripping** — `<think>...</think>` blocks filtered at token level before sentence detector (see #6)
- [x] **Server-agnostic token limits** — sends both `max_tokens` (OpenAI) and `n_predict` (llama.cpp)

### Not tested (deferred to Gateway/integration milestones)
- [ ] Opus decode path (`feed_opus()` → OpusDecoder → RealtimeSTT)
- [ ] PREPARE / barge-in (cancel event, speaker stop/split, spoken/unspoken accounting)
- [ ] PrefillCache (KV cache warming between turns)
- [ ] Custom TurnDetector two-tier logic (harness uses RealtimeSTT's built-in turn detection)
- [ ] Vocabulary rules (model didn't produce triggering words)
- [ ] Tool/delegate tag extraction (model didn't emit `[TOOL:...]` or `[DELEGATE:...]`)
- [ ] gRPC transport (everything runs in-process, no serialization/deserialization)

## Test plan
- [x] `pytest tests/ -v` — 192 passed
- [x] `ruff check . && ruff format --check .` — clean
- [x] `mypy config.py voice/ inference/ --ignore-missing-imports` — clean
- [x] CI passes on push

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: talker/voice/listener.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        await listener.run()  # blocks; cancelled on shutdown
+    """
+
+    def __init__(self, config: TalkerConfig) -> None:
+        self._config = config
+        self._opus = OpusDecoder()
+        self._turn_detector = TurnDetector(config)
+        self._event_queue: asyncio.Queue[kaguya_pb2.ListenerEvent] = asyncio.Queue()
+        self._recorder: AudioToTextRecorder | None = None
+        self._loop: asyncio.AbstractEventLoop | None = None
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/voice/listener.py:43]: deferred

File: talker/inference/llm_client.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+) -> AsyncIterator[str]:
+    """Parse Server-Sent Events with cancellation support.
+
+    Expected format (llama.cpp / LM Studio):
+        data: {"content": "Hello", "stop": false}
+        data: {"content": "", "stop": true}
+        data: [DONE]
+
+    If cancel_event is set, the iterator raises asyncio.CancelledError
+    to abort the stream immediately — no waiting for the next token.
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/inference/llm_client.py]: fixed docstring in [1a3b64b](https://github.com/styin/kaguya/pull/4/commits/1a3b64bc895eb9241bf4b80fd8d86c90c47a07ae)

File: talker/server.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            # Flush remaining buffer on clean completion.
+            # Note: sentence detection boundaries only detect up to the penultimate sentence.
+            if not was_interrupted:
+                remainder = detector.flush()
+                if remainder:
+                    sentence_count += 1
+                    if sentence_count <= self._config.max_response_sentences:
+                        seq = yield from _emit_sentence(
+                            remainder, self._identity, self._speaker, seq
+                        )
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/server.py]: fixed and resolved by [1a3b64b](https://github.com/styin/kaguya/pull/4/commits/1a3b64bc895eb9241bf4b80fd8d86c90c47a07ae)

File: talker/server.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+                sentence = detector.feed(
+                    token
+                )  # Returns `None` until a sentence is detected
+                if sentence is not None:
+                    sentence_count += 1
+
+                    # Emit sentence for post-processing -> TalkerOutput + TTS
+                    seq = yield from _emit_sentence(
+                        sentence, self._identity, self._speaker, seq
+                    )
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/server.py]: fixed and resolved by [1a3b64b](https://github.com/styin/kaguya/pull/4/commits/1a3b64bc895eb9241bf4b80fd8d86c90c47a07ae)

File: talker/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+version = "0.1.0"
+description = "Kaguya Talker Agent — voice listener, LLM pipeline, TTS speaker"
+requires-python = ">=3.11"
+dependencies = [
+    "RealtimeSTT",
+    "RealtimeTTS[kokoro]",
+    "opuslib",
+    "grpcio>=1.60.0",
+    "pydantic>=2.0",
+    "pydantic-settings>=2.0",
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/pyproject.toml:13]: fixed and resolved by [1a3b64b](https://github.com/styin/kaguya/pull/4/commits/1a3b64bc895eb9241bf4b80fd8d86c90c47a07ae)

File: talker/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    "httpx>=0.27.0",
+]
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/pyproject.toml:15]: fixed and resolved by [1a3b64b](https://github.com/styin/kaguya/pull/4/commits/1a3b64bc895eb9241bf4b80fd8d86c90c47a07ae)

File: talker/voice/turn_detector.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            return None
+
+        if silence_ms < self._silence_ms:
+            # Ambiguous zone: use syntactic shape as discriminator.
+            if self._is_syntactically_complete(self._buffer):
+                return self._emit()
+            return None  # Incomplete — keep waiting.
+
+        # Unconditional emit at or beyond SILENCE_THRESHOLD_MS.
+        return self._emit()
```
[the remark this answers is not in the document]
Comment (styin) [reply talker/voice/turn_detector.py:142]: added comment in [1a3b64b](https://github.com/styin/kaguya/pull/4/commits/1a3b64bc895eb9241bf4b80fd8d86c90c47a07ae)
